runtime/src/mocks.rs

```diff
@@ -1,43 +1,241 @@
-//! Mock implementations of `commonware-runtime` traits for focused, upstream testing.
+//! A mock implementation of a channel that implements the Sink and Stream traits.
 
-use crate::{Error, Sink, Stream};
+use crate::{
+    Error,
+    Sink as SinkTrait,
+    Stream as StreamTrait,
+};
 use bytes::Bytes;
-use futures::{channel::mpsc, SinkExt, StreamExt};
+use futures::channel::oneshot;
+use std::{
+    collections::VecDeque,
+    sync::{Arc, Mutex},
+};
 
-/// A mock sink that stores sent messages in a channel.
-pub struct MockSink {
-    pub sender: mpsc::UnboundedSender<Bytes>,
+// A mock channel struct that is used internally by Sink and Stream.
+struct Channel {
+    // Stores the bytes sent by the sink that are not yet read by the stream.
+    buffer: VecDeque<u8>,
+
+    // If the stream is waiting to read bytes, the waiter stores the number of
+    // bytes that the stream is waiting for, as well as the oneshot sender that
+    // the sink uses to send the bytes to the stream directly.
+    waiter: Option<(usize, oneshot::Sender<Bytes>)>,
 }
 
-impl MockSink {
-    /// Create a new `MockSink` and a corresponding receiver.
-    pub fn new() -> (Self, mpsc::UnboundedReceiver<Bytes>) {
-        let (sender, receiver) = mpsc::unbounded();
-        (Self { sender }, receiver)
-    }
+/// Returns an async-safe Sink/Stream pair that share an underlying buffer of bytes.
+pub fn new() -> (Sink, Stream) {
+    let channel = Arc::new(Mutex::new(Channel {
+        buffer: VecDeque::new(),
+        waiter: None,
+    }));
+    (
+        Sink { channel: channel.clone() },
+        Stream { channel },
+    )
 }
 
-impl Sink for MockSink {
-    async fn send(&mut self, msg: Bytes) -> Result<(), Error> {
-        self.sender.send(msg).await.map_err(|_| Error::WriteFailed)
+/// A mock sink that implements the Sink trait.
+pub struct Sink {
+    channel: Arc<Mutex<Channel>>,
+}
+
+impl SinkTrait for Sink {
+    /// Writes the message to the buffer.
```

Comment:
nit: you don't need to comment implementations of a trait (the trait should already have sufficient commentary explaining what it does)